flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecMultiJoin.java

```diff
@@ -0,0 +1,444 @@
+            }
+        }
+        return instantiatedJoinConditions;
+    }
+
+    private StreamOperatorFactory<RowData> createOperatorFactory(
+            ExecNodeConfig config,
+            List<InternalTypeInfo<RowData>> inputTypeInfos,
+            JoinCondition[] joinConditions,
+            JoinKeyExtractor keyExtractor) {
+        List<Long> stateTtls =
+                StateMetadata.getStateTtlForMultiInputOperator(
+                        config, inputTypeInfos.size(), stateMetadataList);
+        long[] stateRetentionTimes = stateTtls.stream().mapToLong(Long::longValue).toArray();
+
+        return new StreamingMultiJoinOperatorFactoryImpl(
+                inputTypeInfos,
+                inputSideSpecs,
+                joinTypes,
+                null, // multiJoinCondition is currently not used
+                stateRetentionTimes,
+                joinConditions,
+                keyExtractor,
+                joinAttributeMap);
+    }
+
+    private Transformation<RowData> createTransformation(
+            List<Transformation<RowData>> inputTransforms,
+            TransformationMetadata metadata,
+            StreamOperatorFactory<RowData> operatorFactory) {
+        if (inputTransforms.isEmpty()) {
+            throw new IllegalStateException("StreamExecMultiJoin requires at least two inputs.");
+        }
+
+        return ExecNodeUtil.createKeyedMultiInputTransformation(
+                inputTransforms,
+                commonJoinKeySelectorsForTransformation,
+                commonJoinKeyType,
+                metadata,
+                operatorFactory,
+                InternalTypeInfo.of(getOutputType()),
+                inputTransforms.get(0).getParallelism(),
+                false);
+    }
+
+    private GeneratedJoinCondition generateJoinConditionForInput(
+            ExecNodeConfig config,
+            ClassLoader classLoader,
+            RexNode joinCondition,
+            List<RowType> inputRowTypes,
+            int inputIndex) {
+        if (inputIndex == 0) {
+            return null;
+        }
+
+        RowType leftType = leftTypeForIndex(inputRowTypes, inputIndex);
+        RowType rightType = inputRowTypes.get(inputIndex);
+
+        return JoinUtil.generateConditionFunction(
+                config, classLoader, joinCondition, leftType, rightType);
+    }
+
+    private RowType leftTypeForIndex(List<RowType> inputRowTypes, int inputIndex) {
+        if (inputIndex <= 0) {
+            throw new IllegalArgumentException(
+                    "Input index must be greater than 0 for accumulated left type calculation");
+        }
+
+        RowType leftType = inputRowTypes.get(0);
+        for (int i = 1; i < inputIndex; i++) {
+            leftType = RowType.of(leftType, leftType);
+        }
+
+        return leftType;
+    }
+
+    /** Serializable factory for creating {@link StreamingMultiJoinOperator}. */
+    private static class StreamingMultiJoinOperatorFactoryImpl
```

Comment:
Factories are send to TaskManagers. That's why they belong into `org.apache.flink.table.runtime` module. Make sure that all data types in this class are serializable.

Reply:
Also double checked and they all implement Serializable if I didn't miss anything